### DISCUSSION TITLE
Base64 validation

### DISCUSSION THREAD

**Discussion by MrShnaider** [laravel/framework#50487] [Q&A]
Is there any way in Laravel to validate base64 encoded files? For example, check that the parameter is base64 encoded, that it is png encoded, or that its maximum size is no more than a few MB? It looks like a common use case, but I didn't find it in the documentation.

**Comment by ghost** [comment]
https://laravel.com/docs/10.x/validation#rule-mimes
https://laravel.com/docs/10.x/validation#rule-max
https://laravel.com/docs/10.x/validation#custom-validation-rules and build your custom validation with base64_decode

**Comment by Raza9798** [comment]
> https://laravel.com/docs/10.x/validation#rule-mimes https://laravel.com/docs/10.x/validation#rule-max https://laravel.com/docs/10.x/validation#custom-validation-rules and build your custom validation with base64_decode

To make a custom validation run the command `php artisan make:rule Base64Image` and implement the function as below
```
<?php

namespace App\Rules;

use Illuminate\Contracts\Validation\Rule;

class Base64Image implements Rule
{
    public function passes($attribute, $value)
    {
        if (!preg_match('/^data:image\/(\w+);base64,/', $value)) {
            return false;
        }

        list($type, $data) = explode(';', $value);
        list(, $data)      = explode(',', $data);

        if (!in_array($type, ['image/png', 'image/jpeg', 'image/gif'])) {
            return false;
        }

        if (strlen(base64_decode($data)) > 2 * 1024 * 1024) {
            return false;
        }

        return true;
    }

    public function message()
    {
        return 'The :attribute must be a valid base64 encoded image (PNG, JPEG, GIF) with a maximum size of 2MB.';
    }
}
```

to call the validation you can try as follows

```
use App\Rules\Base64Image;

public function store(Request $request)
{
    $request->validate([
        'image' => ['required', new Base64Image],
    ]);
 
}
```

**Reply by SolankiJignashu** [reply]
Before
```php
list($type, $data) = explode(';', $value);
``` 
we should remove the `data:` prefix from the base64 string (if present) to avoid incorrect parsing.
```php
if (str_starts_with($value, 'data:')) {
      $value = substr($value, 5);
}
list($type, $data) = explode(';', $value);
list(, $data)      = explode(',', $data);
``` 
This ensures the base64 data is correctly split into `$type` and `$data`

**Reply by shaedrich** [reply]
Instead of specifically images, we should make a general base64 rule where `mimes` can be applied

**Reply by SolankiJignashu** [reply]
Yes. That can be good solution.

**Reply by Raza9798** [reply]
```php
<?php

namespace App\Rules;

use Illuminate\Contracts\Validation\Rule;

class Base64Image implements Rule
{
    public function passes($attribute, $value)
    {
        return preg_match('#^data:image/(png|jpeg|gif);base64,#', $value)
            && ($data = base64_decode(substr($value, strpos($value, ',') + 1), true))
            && strlen($data) <= 2 * 1024 * 1024
            && @getimagesizefromstring($data);
    }

    public function message()
    {
        return 'The :attribute must be a valid base64 encoded PNG, JPEG, or GIF image under 2MB.';
    }
}
```

try me latest validation @shaedrich @SolankiJignashu

**Reply by shaedrich** [reply]
So no jpg (just jpeg) and no other image formats (looking at you webp)?
